src/libxtp/basisset.cc

```diff
@@ -174,70 +138,75 @@ std::vector<Index> NumFuncSubShell(const std::string& shell_type) {
 
 Index NumFuncShell_cartesian(const std::string& shell_type) {
   Index nbf = 0;
+  assert(!shell_type.empty() && "Shelltype must be non empty!");
   // single type shells defined here
-  if (shell_type.length() == 1) {
-    if (shell_type == "S") {
-      nbf = 1;
-    } else if (shell_type == "P") {
-      nbf = 3;
-    } else if (shell_type == "D") {
-      nbf = 6;
-    } else if (shell_type == "F") {
-      nbf = 10;
-    } else if (shell_type == "G") {
-      nbf = 15;
-    } else if (shell_type == "H") {
-      nbf = 21;
-    } else if (shell_type == "I") {
-      nbf = 28;
+  for (const char& t : shell_type) {
+    if (t == 'S') {
+      nbf += 1;
+    } else if (t == 'P') {
+      nbf += 3;
+    } else if (t == 'D') {
+      nbf += 6;
+    } else if (t == 'F') {
+      nbf += 10;
+    } else if (t == 'G') {
+      nbf += 15;
+    } else if (t == 'H') {
+      nbf += 21;
+    } else if (t == 'I') {
+      nbf += 28;
     } else {
       throw std::runtime_error("NumFuncShell_cartesian shell_type not known");
     }
+  }
+  return nbf;
+}
+
+Index OffsetFuncShell_cartesian(const std::string& shell_type) {
+  Index nbf = std::numeric_limits<Index>::max();
+  assert(!shell_type.empty() && "Shelltype must be non empty!");
+  const char t = shell_type[0];
+  if (t == 'S') {
+    nbf = 0;
+  } else if (t == 'P') {
+    nbf = 1;
+  } else if (t == 'D') {
+    nbf = 4;
+  } else if (t == 'F') {
+    nbf = 10;
+  } else if (t == 'G') {
+    nbf = 20;
+  } else if (t == 'H') {
+    nbf = 35;
+  } else if (t == 'I') {
+    nbf = 56;
   } else {
-    // for combined shells, sum over all contributions
-    for (Index i = 0; i < Index(shell_type.length()); ++i) {
-      std::string local_shell = std::string(shell_type, i, 1);
-      nbf += NumFuncShell_cartesian(local_shell);
-    }
+    throw std::runtime_error("OffsetFuncShell_cartesian shell_type not known");
   }
 
   return nbf;
 }
 
-Index OffsetFuncShell_cartesian(const std::string& shell_type) {
-  Index nbf;
-  // single type shells
-  if (shell_type.length() == 1) {
-    if (shell_type == "S") {
-      nbf = 0;
-    } else if (shell_type == "P") {
-      nbf = 1;
-    } else if (shell_type == "D") {
-      nbf = 4;
-    } else if (shell_type == "F") {
-      nbf = 10;
-    } else if (shell_type == "G") {
-      nbf = 20;
-    } else if (shell_type == "H") {
-      nbf = 35;
-    } else if (shell_type == "I") {
-      nbf = 56;
-    } else {
-      throw std::runtime_error(
-          "OffsetFuncShell_cartesian shell_type not known");
-    }
+bool CheckShellType(const std::string& shelltype) {
+  if (shelltype.empty()) {
+    return false;
+  }
+  std::vector<char> allowed_shells = {'S', 'P', 'D', 'F', 'G', 'H', 'I'};
+  std::vector<char>::iterator it =
```

Comment:
**auto** can help here to reduce the *Carpal tunnel syndrome* ;)

Reply:
:) I prefer the explicit statements a bit more to be honest. I see your point though.